ansible_base/authenticator_plugins/saml.py

```diff
@@ -0,0 +1,191 @@
+        allow_null=False,
+        help_text=_("The field in the assertion which represents the users email."),
+    )
+    IDP_ATTR_USERNAME = serializers.CharField(
+        help_text=_("The field in the assertion which represents the users username."),
+    )
+    IDP_ATTR_LAST_NAME = serializers.CharField(
+        allow_null=False,
+        help_text=_("The field in the assertion which represents the users last name."),
+    )
+    IDP_ATTR_FIRST_NAME = serializers.CharField(
+        allow_null=False,
+        help_text=_("The field in the assertion which represents the users first name."),
+    )
+    IDP_ATTR_USER_PERMANENT_ID = serializers.CharField(
+        help_text=_("The field in the assertion which represents the users permanent id (overrides IDP_ATTR_USERNAME)"),
+    )
+
+    def validate(self, attrs):
+        # attrs is only the data in the configuration field
+        errors = {}
+        # pull the cert_info out of the existing object (if we have one)
+        cert_info = {
+            "SP_PRIVATE_KEY": getattr(self.instance, 'configuration', {}).get('SP_PRIVATE_KEY', None),
+            "SP_PUBLIC_CERT": getattr(self.instance, 'configuration', {}).get('SP_PUBLIC_CERT', None),
+        }
+
+        # Now get the cert_info out of the passed in attrs (if there is any)
+        for cert_type in cert_info.keys():
+            item = attrs.get(cert_type, None)
+            if item and item == ENCRYPTED_STRING and not self.instance:
+                # Catch a case where we got input but it was ENCRYPTED and we don't have an object yet
+                errors[cert_type] = f"Can not be {ENCRYPTED_STRING} on creation"
+            elif item and item != ENCRYPTED_STRING:
+                # We got an input form the attrs so let that override whatever was in the object
+                cert_info[cert_type] = item
+            # If we didn't get an input or we got ENCRYPTED_STRING but there is an item, we will just use whatever we got from the item
+
+        # If we made it here the cert_info has one of three things:
+        #  * None (error state or not passed in on PUT)
+        #  * The existing value from the instance
+        #  * A new value
+
+        # Now validate that we can load the cert and key and that they match.
+        # Technically, we are also doing this on save even if both values came from the existing instance
+        # so there is an inefficiency here but it should be trivial
+        try:
+            validate_cert_with_key(cert_info['SP_PUBLIC_CERT'], cert_info['SP_PRIVATE_KEY'])
+        except ValidationError as e:
+            errors['SP_PRIVATE_KEY'] = e
+
+        idp_data = attrs.get('ENABLED_IDPS', {}).get(idp_string, {})
+        # TODO: Check to make sure either ID or Perminant ID is set
+        if not idp_data['attr_user_permanent_id'] and not idp_data['attr_username']:
+            errors['IDP_ATTR_USERNAME'] = "Either IDP_ATTR_USERNAME or IDP_ATTR_USER_PERMANENT_ID needs to be set"
```

Comment:
Both the `IDP_ATTR_USERNAME` and `IDP_ATTR_USER_PERMANENT_ID` fields default to required, so I think this is dead code currently?

Should one or both of those be optional?

Reply:
They shouldn't have been, fixed.